tee/prototype/epochproc.go

```diff
@@ -0,0 +1,397 @@
+		// extract deposits, adjust epoch's balances.
+		if err := e.applyEpochDeposit(e.params.Contract, depEpoch, depLogs); err != nil {
+			return nil, fmt.Errorf("adjusting Epoch %v Balances: %w", depEpoch.Number, err)
+		}
+		// extract exits, adjust epoch's balances.
+		exits, err := e.applyEpochExit(exitEpoch, exLogs)
+		if err != nil {
+			return nil, fmt.Errorf("handling exit of Epoch %v: %w", exitEpoch.Number, err)
+		}
+		exiters = append(exiters, exits...)
+	}
+	return exiters, nil
+}
+
+type noPredLogs = []*types.Log
+type predLogs = []*types.Log
+
+// partition partitions a given slice of `*types.Log` into a slice where `pred`
+// holds and vice versa.
+func partition(ls []*types.Log, pred func(l *types.Log) bool) (predLogs, noPredLogs) {
+	var depLogs, exLogs []*types.Log
+	for _, l := range ls {
+		if pred(l) {
+			depLogs = append(depLogs, l)
+		} else {
+			exLogs = append(exLogs, l)
+		}
+	}
+	return depLogs, exLogs
+}
+
+// applyEpochDeposit adjusts `e.balances` according to the deposits done in
+// the given block.
+func (e *Enclave) applyEpochDeposit(contract common.Address, ep *Epoch, depLogs []*types.Log) error {
+	for _, depLog := range depLogs {
+		deposit, err := parseDepEvent(depLog)
+		if err != nil {
+			return fmt.Errorf("parsing withdraw event: %w", err)
+		}
+		if accBal, ok := ep.balances[deposit.Account]; ok {
+			accBal.Value.Add(accBal.Value, deposit.Value)
+		} else {
+			ep.balances[deposit.Account] = Bal{
+				0,
+				new(big.Int).Set(deposit.Value),
+			}
+		}
+		if deposit.Epoch != ep.Number {
+			return fmt.Errorf("deposit-event Epoch %v != current deposit Epoch %v",
+				deposit.Epoch, ep.Number)
+		}
+
+		depProof, err := e.generateDepositProof(ep, deposit.Account)
+		if err != nil {
+			return fmt.Errorf("generating deposit proof: %w", err)
+		}
+		e.depositProofCache[deposit.Account] = depProof
+	}
+	return nil
+}
+
+// exitersSet is the set of exiting participants.
+type exitersSet []common.Address
+
+// applyEpochExit handles the exit phase of given Epoch.
+func (e *Enclave) applyEpochExit(ep *Epoch, exLogs []*types.Log) (exitersSet, error) {
+	exiters := make(exitersSet, 0)
+	for _, exLog := range exLogs {
+		exit, err := parseExitEvent(exLog)
+		if err != nil {
+			return nil, fmt.Errorf("parsing exiting event: %w", err)
+		}
+		if exit.Epoch != ep.Number {
+			return nil, fmt.Errorf("exit-event Epoch %v != current exit Epoch %v",
+				exit.Epoch, ep.Number)
+		}
+		// only full exits supported currently
+		accBal := ep.balances[exit.Account].Value
+		if accBal.Cmp(exit.Value) != 0 {
+			return nil, fmt.Errorf("unexpected partial exit for %v, expected %v", exit.Value, accBal)
+		}
+		accBal.SetUint64(0)
+		delete(ep.balances, exit.Account)
+		exiters = append(exiters, exit.Account)
+	}
+
+	return exiters, nil
+}
+
+// applyEpochTx adjusts `e.balances` according to given transactions.
+func (e *Enclave) applyEpochTx(ep *Epoch, txs []*tee.Transaction, bpc balProofCache) (balProofCache, error) {
```

Comment:
why return the cache? also note that in the last line, you just return `nil, nil`.

Reply:
see [comment](https://github.com/perun-network/erdstall/pull/3#discussion_r507686815).

and good catch.